[serve] Add Autoscaling Guide (#39219)
ray-project/ray: doc/source/serve/autoscaling-guide.md

The results of the load test are as follows:

|  |  |  |
| -------- | --- | ------- |
| Replicas | <img src="https://raw.githubusercontent.com/ray-project/images/master/docs/serve/autoscaling-guide/resnet50_replicas.png" alt="replicas" width="600"/> |
| QPS | <img src="https://raw.githubusercontent.com/ray-project/images/master/docs/serve/autoscaling-guide/resnet50_rps.svg" alt="qps"/> |
| P50 Latency | <img src="https://raw.githubusercontent.com/ray-project/images/master/docs/serve/autoscaling-guide/resnet50_latency.svg" alt="latency"/> |

Notice the following:
- Each Locust user constantly sends a single request and waits for a response. As a result, the number of autoscaled replicas closely matches the number of Locust users over time as Serve attempts to satisfy the `target_num_ongoing_requests_per_replica=1` setting.
- The throughput of the system increases with the number of users and replicas.
- The latency briefly spikes when traffic increases, but otherwise stays relatively steady. The biggest latency spike happens at the beginning of the test, because the deployment starts with 0 replicas.

> ## Model Composition Example

Author: This can be de-capitalized!

**upscale_delay_s[default_value=30.0]**: How long the cluster needs to wait before scaling up replicas. This controls the frequency of upscale decisions.

**upscale_smoothing_factor[default_value=1.0]**: The multiplicative factor to speed up or slow down each upscaling decision. For example, when the application has high traffic volume in a short period of time, you can increase `upscale_smoothing_factor` to scale up the resource quickly. You can think of this as a "gain" factor to amplify the response of the autoscaling algorithm.

**downscale_smoothing_factor[default_value=1.0]**: The multiplicative factor to speed up or slow down each downscaling decision. For example, if you want your application to be less sensitive to drops in traffic and scale down more conservatively, you can decrease `downscale_smoothing_factor` to slow down the pace of downscaling.

> **metrics_interval_s[default_value=10]**: This controls how often each replica sends reports on current ongoing requests to the autoscaler. Note that the autoscaler cannot make new decisions if it does not receive updated metrics, so you most likely want to set `metrics_interval_s` to a value that is at most the upscale and downscale delay values. For instance, if you set `upscale_delay_s = 3`, but keep `metrics_interval_s = 10`, the autoscaler will only upscale roughly every 10 seconds.

Reviewer: ```suggestion
**metrics_interval_s[default_value=10]**: This controls how often each replica sends reports on current ongoing requests to the autoscaler. Note that the autoscaler can't make new decisions if it doesn't receive updated metrics, so you most likely want to set `metrics_interval_s` to a value that is at most the upscale and downscale delay values. For instance, if you set `upscale_delay_s = 3`, but keep `metrics_interval_s = 10`, the autoscaler only upscales roughly every 10 seconds.
```

> **look_back_period_s[default_value=30]**: This controls how often each replica sends reports on current ongoing requests to the autoscaler. Note that the autoscaler cannot make new decisions if it does not receive updated metrics, so you most likely want to set `metrics_interval_s` to a value that is at most the upscale and downscale delay values. For instance, if you set `upscale_delay_s = 3`, but keep `metrics_interval_s = 10`, the autoscaler will only upscale roughly every 10 seconds.

Reviewer: ```suggestion
**look_back_period_s[default_value=30]**: This parameter controls how often each replica sends reports on current ongoing requests to the autoscaler. Note that the autoscaler can't make new decisions if it doesn't receive updated metrics, so you most likely want to set `metrics_interval_s` to a value that is at most the upscale and downscale delay values. For instance, if you set `upscale_delay_s = 3`, but keep `metrics_interval_s = 10`, the autoscaler only upscales roughly every 10 seconds.
```

> ## Troubleshooting Guide

Reviewer: ```suggestion
## Troubleshooting guide
```

> ### Number of Autoscaled Replicas not Stable

Reviewer: ```suggestion
### Unstable number of autoscaled replicas
```

If you are seeing that the number of replicas in your deployment keeps oscillating even though the traffic is relatively stable, here are some things you can try:

* Set a smaller `upscale_smoothing_factor` and `downscale_smoothing_factor`.
* Set a `look_back_period_s` value that matches the rest of the autoscaling config. For longer upscale and downscale delay values, a longer look back period can likely help stabilize the replica graph, but for shorter upscale and downscale delay values, a shorter look back period may be more appropriate. For instance, the following replica graphs show how a deployment with `upscale_delay_s = 3` works with a longer vs shorter look back period.

| `look_back_period_s = 30` | `look_back_period_s = 3` |
| ------------------------------------------------ | ----------------------------------------------- |
| ![look-back-before](https://raw.githubusercontent.com/ray-project/images/master/docs/serve/autoscaling-guide/look_back_period_before.png) | ![look-back-after](https://raw.githubusercontent.com/ray-project/images/master/docs/serve/autoscaling-guide/look_back_period_after.png) |

### High spikes in latency during bursts of traffic

> If you expect your application to receive bursty traffic, and at the same time want the deployments to scale down in periods of inactivity, you are likely concerned about how quickly the deployment can scale up and respond to bursts of traffic. While an increase in latency initially during a burst in traffic may be unavoidable, here are some things you can try to improve latency during bursts of traffic.

Reviewer: ```suggestion
If you expect your application to receive bursty traffic, and at the same time want the deployments to scale down in periods of inactivity, you are likely concerned about how quickly the deployment can scale up and respond to bursts of traffic. While an increase in latency initially during a burst in traffic may be unavoidable, you can try the following to improve latency during bursts of traffic:
```

> * Set a lower `upscale_delay_s`. The autoscaler will always wait `upscale_delay_s` seconds before making a decision to upscale, so lowering this delay allows the autoscaler to react more quickly to changes, especially bursts, of traffic.

Reviewer: ```suggestion
* Set a lower `upscale_delay_s`. The autoscaler always waits `upscale_delay_s` seconds before making a decision to upscale, so lowering this delay allows the autoscaler to react more quickly to changes, especially bursts, of traffic.
```

* Set a larger `upscale_smoothing_factor`. If `upscale_smoothing_factor > 1`, then the autoscaler will scale up more aggressively than normal. This can allow your deployment to be more sensitive to bursts of traffic.

> * Lower the `metric_interval_s`. we recommend to always set `metric_interval_s` to be less than or equal to `upscale_delay_s`, otherwise upscaling will be delayed because the autoscaler doesn't receive fresh information often enough.

Reviewer: ```suggestion
* Lower the `metric_interval_s`. Always set `metric_interval_s` to be less than or equal to `upscale_delay_s`, otherwise upscaling is delayed because the autoscaler doesn't receive fresh information often enough.
```

> * Set a lower `max_concurrent_queries`. If `max_concurrent_queries` is too high relative to `target_num_ongoing_requests_per_replica`, then when traffic increases, most or all of the requests might be assigned to the existing replicas before the new replicas are started. This can lead to very high latencies during upscale.

Reviewer: ```suggestion
* Set a lower `max_concurrent_queries`. If `max_concurrent_queries` is too high relative to `target_num_ongoing_requests_per_replica`, then when traffic increases, most or all of the requests might be assigned to the existing replicas before the new replicas are started. This setting can lead to very high latencies during upscale.
```

> ### Want deployment to scale down more conservatively

Reviewer: Since this is a troubleshooting guide, the titles are describing problems vs solutions.

You may observe that deployments are scaling down too quickly for your liking. Instead, you may want the downscaling to be much more conservative to maximize the availability of your service.

> * Set a longer `downscale_delay_s`. The autoscaler will always wait `downscale_delay_s` seconds before making a decision to downscale, so by increasing this number, your system will have a longer "grace period" after traffic drops before the autoscaler starts to remove replicas.

Reviewer: ```suggestion
* Set a longer `downscale_delay_s`. The autoscaler always waits `downscale_delay_s` seconds before making a decision to downscale, so by increasing this number, your system has a longer "grace period" after traffic drops before the autoscaler starts to remove replicas.
```

> * Set a smaller `downscale_smoothing_factor`. If `downscale_smoothing_factor < 1`, then the autoscaler will remove *less replicas* than what it thinks it should remove to achieve the target number of ongoing requests. In other words, the autoscaler makes more conservative downscaling decisions.

Reviewer: ```suggestion
* Set a smaller `downscale_smoothing_factor`. If `downscale_smoothing_factor < 1`, then the autoscaler removes *less replicas* than what it thinks it should remove to achieve the target number of ongoing requests. In other words, the autoscaler makes more conservative downscaling decisions.
```

(serve-autoscaling)=

# Autoscaling

## Scaling horizontally with `num_replicas`

> One way to scale out your deployment without using autoscaling is by increasing the number of replicas. Each deployment consists of one or more [replicas](serve-architecture-high-level-view).

Reviewer 2: it is, you can set your min_replicas to `0` and then it will scale up as soon as any requests are seen

The number of replicas is specified by the `num_replicas` field in the deployment options.
By default, `num_replicas` is 1.

```{literalinclude} doc_code/managing_deployments.py
:start-after: __scaling_out_start__
:end-before: __scaling_out_end__
:language: python
```

## Autoscaling Important Concepts

> Instead of setting a fixed number of replicas for a deployment, you can configure a deployment to autoscale based on incoming traffic. The Serve autoscaler adjusts to traffic spikes by observing queue sizes and making scaling decisions to add or remove replicas. To configure it, you can set the `autoscaling_config` field in deployment options.

Reviewer 2: please cross-link to the API ref whenever referencing APIs/fields like this (in this case, link to `@serve.deployment`)

Author: I added `AutoscalingConfig` to the api refs and linked to it!

> Instead of setting a fixed number of replicas for a deployment, you can configure a deployment to autoscale based on incoming traffic. The Serve autoscaler adjusts to traffic spikes by observing queue sizes and making scaling decisions to add or remove replicas. To configure it, you can set the `autoscaling_config` field in deployment options.

Reviewer 3: ```suggestion
Instead of setting a fixed number of replicas for a deployment, you can configure a deployment to autoscale based on incoming traffic. The Serve autoscaler reacts to traffic spikes by monitoring queue sizes and making scaling decisions to add or remove replicas. Configure it by setting the `autoscaling_config` field in deployment options.
```

:::{note}
The Ray Serve Autoscaler is an application-level autoscaler that sits on top of the [Ray Autoscaler](cluster-index).
Concretely, this means that the Ray Serve autoscaler asks Ray to start a number of replica actors based on the request demand.

> If the Ray Autoscaler determines there aren't enough available CPUs to place these actors, it responds by requesting more Ray nodes.

Reviewer 3: ```suggestion
If the Ray Autoscaler determines there aren't enough available resources (e.g. CPUs, GPUs, etc.) to place these actors, it responds by requesting more Ray nodes.
```

The underlying cloud provider will then respond by adding more nodes.

> Similarly, when Ray Serve scales down and terminates some replica actors, it will try to do so in a way that results in the most nodes having no Ray actors or tasks running on them, at which point the Ray autoscaler will remove those nodes.

Reviewer 2: ```suggestion
Similarly, when Ray Serve scales down and terminates replica actors, it will attempt to make as many nodes idle as possible so the Ray autoscaler can remove them.
```

:::

To learn about the architecture underlying Ray Serve Autoscaling, see {ref}`serve-autoscaling-architecture`.

Before diving into an example, we will go over some foundational concepts that Serve autoscaling is built on top of.

### Define the Steady State of your System

To define what the steady state of your deployments should be like, you need to set values for `target_num_ongoing_requests_per_replica` and `max_concurrent_queries`.

Serve will scale the number of replicas for a deployment up or down based on the average number of ongoing requests per replica. More specifically, Serve will compare the *actual* average number of ongoing requests per replica with the target value you set in the autoscaling config and make upscale or downscale decisions from that. The target value is set by `target_num_ongoing_requests_per_replica`, and Serve will try to make sure that each replica has roughly that number

> of requests being processed and waiting in the queue. We recommend you benchmark your application code and set this number based on end to end latency objective.  There is also a maximum limit for the queue that is respected by proxies when assigning requests to replicas, which is defined by `max_concurrent_queries`.

Reviewer 2: ```suggestion
of requests being processed and waiting in the queue. We recommend you benchmark your application code and set this number based on an end to end latency objective.  There is also a maximum limit for the queue that is respected by proxies when assigning requests to replicas, which is defined by `max_concurrent_queries`.
```

Reviewer 2: given we mention `max_concurrent_queries`, we should also specify that `target_num_ongoing_requests_per_replica` should never be higher than it (or you won't scale up)

Author: Added!

:::{note}
- It is always recommended to load test your workloads. For example, if the use case is latency sensitive, you can lower the `target_num_ongoing_requests_per_replica` number to maintain high performance.

> - Internally, the autoscaler will decide to scale up or down by comparing `target_num_ongoing_requests_per_replica` to the number of `RUNNING` and `PENDING` tasks on each replica.

Reviewer 2: ```suggestion
- Internally, the autoscaler will decide to scale up or down by comparing `target_num_ongoing_requests_per_replica` to the number of requests running and pending on each replica.
```

`RUNNING` and `PENDING` are not publicly-exposed states so this may be confusing

- `target_num_ongoing_requests_per_replica` is only a target value used for autoscaling (not a hard limit), the real ongoing requests number can be higher than the config.
:::

As an example, suppose you have two replicas of a synchronous deployment that has 100ms latency, serving a traffic load of 30 QPS. Then requests will be assigned to replicas faster than the replicas can finish processing them; more and more requests will be queued up at the replica (these are considered "ongoing requests") as time goes on, and so the average number of ongoing requests at each replica will steadily increase. If you set `target_num_ongoing_requests_per_replica = 1`, Serve will detect a higher-than-desired number of ongoing requests per replica, and add more replicas. At 3 replicas, your system would be able to process 30 QPS with 1 ongoing request per replica on average.

> For **`target_num_ongoing_requests_per_replica`**, typically we suggest the following.

Author: The default values are included in the `Autoscaling Config Parameters` section.

* Set `target_num_ongoing_requests_per_replica = 1` if your deployment is synchronous.
* Set `target_num_ongoing_requests_per_replica > 1` if your deployment is asynchronous.

For **`max_concurrent_queries`**, typically we suggest the following.
* Set `max_concurrent_queries` to a value between 2 and 5 if the target value is equal to 1.
* Set `max_concurrent_queries` to ~20 to 50% higher than the target value if the target value is greater than 1.

### Define Upper and Lower Autoscaling Limits

> <!-- Now that you understand how Serve autoscales your deployment based on ongoing requests at each replica, you can set the upper and lower autoscaling bounds. Given you have a range you expect traffic to fall in, you can map that to  -->

Reviewer 2: intentionally commented out?

To use autoscaling, you need to define the minimum and maximum number of resources allowed for your system.

> * **`min_replicas`**: This is the minimum number of replicas for the deployment. If you want to ensure your system can deal with a certain level of traffic at all times, you can set `min_replicas` to a positive number. On the other hand, if you anticipate periods of no traffic and want to scale to zero to save costs, you can set `min_replicas = 0`.

Reviewer 2: somewhere we should caution that `min_replicas=0` will cause higher tail latencies (not sure if this is the right place)

Author: Added!

> * **`min_replicas`**: This is the minimum number of replicas for the deployment. If you want to ensure your system can deal with a certain level of traffic at all times, you can set `min_replicas` to a positive number. On the other hand, if you anticipate periods of no traffic and want to scale to zero to save costs, you can set `min_replicas = 0`.

Reviewer 3: We should mention that the controller and the head node will always be running, even if all the replicas have scaled to zero.

> * **`max_replicas`**: This is the maximum number of replicas for the deployment. It must be greater than or equal to `min_replicas`. You can determine a reasonable `max_replicas` value from what you expect typical traffic to be like.

Reviewer 2: ```suggestion
* **`max_replicas`**: This is the maximum number of replicas for the deployment. It must be greater than or equal to `min_replicas`.
```
the removed sentence is very low information

> * **`initial_replicas`**: This is the number of replicas that the deployment starts with right after deployment.

Reviewer 2: ```suggestion
* **`initial_replicas`**: This is the number of replicas that will be started initially for the deployment.
```
avoid "deployment" word overloading

> Note that determining values for `min_replicas` and `max_replicas` depends on what you set as `target_num_ongoing_requests_per_replica`.

Reviewer 3: This is a bit vague. Could you elaborate on how they're related?

Author: Agreed it is a bit vague, but not sure how to elaborate in a way that will add useful information, so I've removed this sentence!

### [Optional] Define how the System reacts to changing traffic

> Given a steady stream of traffic and appropriately configured `min_replicas` and `max_replicas`, the steady state of your system is essentially fixed for a chosen configuration value for `target_num_ongoing_requests_per_replica`. Before reaching steady state, however, your system will be reacting to traffic shifts. How you want your system to react to changes in traffic determines how you want to set the remaining autoscaling configurations. We will go over two of the most commonly used parameters here.

Reviewer 3: ```suggestion
Given a steady stream of traffic and appropriately configured `min_replicas` and `max_replicas`, the steady state of your system is essentially fixed for a chosen configuration value for `target_num_ongoing_requests_per_replica`. Before reaching steady state, however, your system will be reacting to traffic shifts. How you want your system to react to changes in traffic determines how you want to set the remaining autoscaling configurations. We will go over two more commonly used parameters here.
```

```yaml
applications:
  - name: default
    import_path: resnet:app
    deployments:
    - name: Model
      max_concurrent_queries: 10
      autoscaling_config:
        target_num_ongoing_requests_per_replica: 1
        min_replicas: 0
        initial_replicas: 0
        max_replicas: 200
        upscale_delay_s: 30
        downscale_delay_s: 600
```

We use [Locust](https://locust.io/) to run a load test against this application. The Locust load test will run a certain number of "users" that ping the Resnet50 service, where each user has a [constant wait time](https://docs.locust.io/en/stable/writing-a-locustfile.html#wait-time-attribute) of 0. This means each user will (repeatedly) send a request, wait for a response, then immediately send the next request. The number of users running over time is shown in the following graph:

![users](https://raw.githubusercontent.com/ray-project/images/master/docs/serve/autoscaling-guide/resnet50_users.png)

The results of the load test are as follows:

::::{tab-set}
:::{tab-item} Replicas

> <img src="https://raw.githubusercontent.com/ray-project/images/master/docs/serve/autoscaling-guide/resnet50_replicas.png" alt="replicas" width="400"/>

Reviewer 3: Is there any way to make the graphs a bit bigger? It's hard to see the axes.

Author: I've made the images bigger!

:::

:::{tab-item} QPS
<img src="https://raw.githubusercontent.com/ray-project/images/master/docs/serve/autoscaling-guide/resnet50_rps.svg" alt="qps" width="400"/>
:::

:::{tab-item} Latency
<img src="https://raw.githubusercontent.com/ray-project/images/master/docs/serve/autoscaling-guide/resnet50_latency.svg" alt="latency" width="400"/>
:::
::::

The results of the load test are as follows:

|  |  |  |

> | -------- | --- | ------- |

Author: This is a table without headers

> | -------- | --- | ------- |

Reviewer 4: Yep, but it's duplicated with the tabs. Looks like this 
<img width="1424" alt="image" src="https://github.com/ray-project/ray/assets/7553988/f3856ec5-f009-448b-b522-9f599e1d4971">

| Replicas | <img src="https://raw.githubusercontent.com/ray-project/images/master/docs/serve/autoscaling-guide/resnet50_replicas.png" alt="replicas" width="450"/> |
| QPS | <img src="https://raw.githubusercontent.com/ray-project/images/master/docs/serve/autoscaling-guide/resnet50_rps.svg" alt="qps" width="450"/> |
| P50 Latency | <img src="https://raw.githubusercontent.com/ray-project/images/master/docs/serve/autoscaling-guide/resnet50_latency.svg" alt="latency" width="450"/> |

> <!-- | ![replicas](https://raw.githubusercontent.com/ray-project/images/master/docs/serve/autoscaling-guide/resnet50_replicas.png){width=200px} | ![qps](https://raw.githubusercontent.com/ray-project/images/master/docs/serve/autoscaling-guide/resnet50_rps.png){width=200px} | ![latency](https://raw.githubusercontent.com/ray-project/images/master/docs/serve/autoscaling-guide/resnet50_latency.svg){width=200px} | -->

Reviewer 3: Remove commented lines once you no longer need them.

> Notice that the number of autoscaled replicas closely matches the number of Locust users over time. This makes sense because you can simplify it to one user is mapped to one deployment replica, and the user will continuously send the right amount of traffic to keep one replica busy. As expected, the throughput of the system also increases with the number of users and replicas. Lastly, the latency sees some spikes when the traffic increases, but otherwise stays relatively steady. Notice that the biggest latency spike happens at the beginning of the test, because the deployment starts with 0 replicas.

Reviewer 3: ```suggestion
Note the following:
- Each Locust user constantly sends a single request and waits for a response. As a result, the number of autoscaled replicas closely matches the number of Locust users over time as Serve attempts to satisfy the `target_num_ongoing_requests_per_replica=1` setting.
- The throughput of the system increases with the number of users and replicas.
- The latency briefly spikes when traffic increases, but otherwise stays relatively steady. Note that the biggest latency spike happens at the beginning of the test, because the deployment starts with 0 replicas.
```

## Model Composition Example

> It is less straightforward to determine an autoscaling configuration for a multi-model application because different deployments will autoscale differently. Every workload will have a different latency and differing levels of concurrency. As a result, it takes experimentation to find the right autoscaling config for a model-composition application.

Reviewer 3: ```suggestion
Determining the autoscaling configuration for a multi-model application requires understanding each deployment's scaling requirements. Every deployment has a different latency and differing levels of concurrency. As a result, it takes experimentation to find the right autoscaling config for a model-composition application.
```

We will go through an example of a simple application with three deployments composed together to build some intuition about multi-model autoscaling. Suppose we have deployments:

> * `HeavyLoad`: A mock 200ms workload with high CPU usage.

Reviewer 3: [Nit] Can we rename:
* `HeavyLoad` -> `Slow`
* `LightLoad` -> `Fast`

That way we can say "the `Slow` deployment" or "the `Fast` deployment" whenever we talk about them.

Open to keeping it the same or renaming to something else though.

Author: That's a good idea, except the images need the deployment name labels for clarity, so I would have to rerun the tests to get the new images 😅 for the sake of getting this cherry picked ASAP, leaning towards keeping the deployment names this time!

Reviewer 3: Good call, let's keep it the same.

* `LightLoad`: A mock 100ms workload with high CPU usage.
* `Driver`: A driver deployment that fans out to the `HeavyLoad` and `LightLoad` deployments and aggregates the two outputs.

### Attempt 1: One `Driver` replica

Let's start off with the following deployment configurations. Since the driver deployment has low CPU usage and is only asynchronously making calls to the downstream deployments, it's reasonable to allocate one fixed `Driver` replica.

::::{tab-set}

:::{tab-item} Driver

```yaml
- name: Driver
  num_replicas: 1
  max_concurrent_queries: 200
```

:::

:::{tab-item} HeavyLoad

```yaml
- name: HeavyLoad
  max_concurrent_queries: 3
  autoscaling_config:
    target_num_ongoing_requests_per_replica: 1
    min_replicas: 0
    initial_replicas: 0
    max_replicas: 200
    upscale_delay_s: 3
    downscale_delay_s: 60
    upscale_smoothing_factor: 0.3
    downscale_smoothing_factor: 0.3
    metrics_interval_s: 2

>     look_pack_period_s: 10

Reviewer 3: We never mention these settings until now. Can we either define them or add a note with a link to where their definitions are?

Author: I added a reference to the section below!

```

:::

:::{tab-item} LightLoad

```yaml
- name: LightLoad
  max_concurrent_queries: 3
  autoscaling_config:
    target_num_ongoing_requests_per_replica: 1
    min_replicas: 0
    initial_replicas: 0
    max_replicas: 200
    upscale_delay_s: 3
    downscale_delay_s: 60
    upscale_smoothing_factor: 0.3
    downscale_smoothing_factor: 0.3
    metrics_interval_s: 2
    look_pack_period_s: 10
```

:::

::::

Running the same Locust load test as we did for the Resnet workload, we get the following results:

| | |
| ----------------------- | ---------------------- |

> | HeavyLoad and LightLoad No. Replicas | <img src="https://raw.githubusercontent.com/ray-project/images/master/docs/serve/autoscaling-guide/model_composition_replicas.png" alt="comp" width="500" /> |

Reviewer 3: ```suggestion
| HeavyLoad and LightLoad Number of Replicas | <img src="https://raw.githubusercontent.com/ray-project/images/master/docs/serve/autoscaling-guide/model_composition_replicas.png" alt="comp" width="500" /> |
```

As one might expect, the number of autoscaled `LightLoad` replicas is roughly half that of autoscaled `HeavyLoad` replicas. Although the same number of requests per second are getting sent to both deployments, `LightLoad` replicas can process twice as many requests per second as `HeavyLoad` replicas can, and so the deployment should need half as many replicas to deal with the same traffic load.

However, the service latency rises to 400+ ms when the number of Locust users increased to 100.

| P50 Latency | QPS |
| ------- | --- |
| ![comp_latency](https://raw.githubusercontent.com/ray-project/images/master/docs/serve/autoscaling-guide/model_comp_latency.svg) | ![comp_rps](https://raw.githubusercontent.com/ray-project/images/master/docs/serve/autoscaling-guide/model_comp_rps.svg) |

> Note that the number of Locust users should roughly match the number of `HeavyLoad` replicas to adequately serve the Locust traffic load. However, when the number of Locust users increased to 100, the `HeavyLoad` deployment struggled to get to 100 replicas, and instead only got to 65 replicas. A suspicion arises when we look at the per-deployment latencies. While `HeavyLoad` and `LightLoad` latencies stayed steady at 200ms and 100ms, `Driver` latencies rose from 220 to 400+ ms. This hints that hundreds of QPS may be harming the `Driver` replica's asynchronous event loop performance.

Reviewer 3: ```suggestion
Note that the number of `HeavyLoad` replicas should roughly match the number of Locust users to adequately serve the Locust traffic. However, when the number of Locust users increased to 100, the `HeavyLoad` deployment failed to scale to 100 replicas, and instead only reached 65 replicas. The per-deployment latencies reveal the root cause. While `HeavyLoad` and `LightLoad` latencies stayed steady at 200ms and 100ms, `Driver` latencies rose from 220 to 400+ ms. This suggests that high Locust workload may be overwhelming the `Driver` replica, potentially by impacting the asynchronous event loop's performance.
```

    min_replicas: 1
    initial_replicas: 1
    max_replicas: 10
    upscale_delay_s: 3
    downscale_delay_s: 60
    upscale_smoothing_factor: 0.3
    downscale_smoothing_factor: 0.3
    metrics_interval_s: 2
    look_pack_period_s: 10
```

:::

:::{tab-item} HeavyLoad

```yaml
- name: HeavyLoad
  max_concurrent_queries: 3
  autoscaling_config:
    target_num_ongoing_requests_per_replica: 1
    min_replicas: 0
    initial_replicas: 0
    max_replicas: 200
    upscale_delay_s: 3
    downscale_delay_s: 60
    upscale_smoothing_factor: 0.3
    downscale_smoothing_factor: 0.3
    metrics_interval_s: 2
    look_pack_period_s: 10
```

:::

:::{tab-item} LightLoad

```yaml
- name: LightLoad
  max_concurrent_queries: 3
  autoscaling_config:
    target_num_ongoing_requests_per_replica: 1
    min_replicas: 0
    initial_replicas: 0
    max_replicas: 200
    upscale_delay_s: 3
    downscale_delay_s: 60
    upscale_smoothing_factor: 0.3
    downscale_smoothing_factor: 0.3
    metrics_interval_s: 2
    look_pack_period_s: 10
```

:::
::::

Running the same Locust load test again, we get the following results:

| | |
| ------------------------------------ | ------------------- |

> | HeavyLoad and LightLoad No. Replicas | <img src="https://raw.githubusercontent.com/ray-project/images/master/docs/serve/autoscaling-guide/model_composition_improved_replicas.png" alt="heavy" width="500"/> |

Reviewer 3: Can we switch `no.` to `number` throughout the tables? I think there's enough space to not need the abbreviation.

| Driver No. Replicas | <img src="https://raw.githubusercontent.com/ray-project/images/master/docs/serve/autoscaling-guide/model_composition_improved_driver_replicas.png" alt="driver" width="500"/>

> With up to 6 `Driver` deployments to receive and distribute the incoming requests, the `HeavyLoad` deployment is able to scale up to 90+ replicas, and `LightLoad` up to 47 replicas. This then helps the application latency to stay consistent as the traffic load increases.

Reviewer 3: ```suggestion
With up to 6 `Driver` deployments to receive and distribute the incoming requests, the `HeavyLoad` deployment successfully scales up to 90+ replicas, and `LightLoad` up to 47 replicas. This helps the application latency stay consistent as the traffic load increases.
```

| Improved P50 Latency | Improved RPS |
| ---------------- | ------------ |
| ![comp_latency](https://raw.githubusercontent.com/ray-project/images/master/docs/serve/autoscaling-guide/model_composition_improved_latency.svg) | ![comp_latency](https://raw.githubusercontent.com/ray-project/images/master/docs/serve/autoscaling-guide/model_comp_improved_rps.svg) |

## Autoscaling Config Parameters

<!-- ```{literalinclude} doc_code/managing_deployments.py
:start-after: __autoscaling_start__
:end-before: __autoscaling_end__
:language: python

> ``` -->

Reviewer 3: Remove commented docs once it's not longer needed.

**min_replicas[default_value=1]**: The minimum number of replicas for the deployment. Ray Serve allows this to be set to 0, which means there can be 0 replicas in periods of inactivity. When you start sending traffic, the deployment will scale up, and there will be a cold start time as Serve waits for replicas to be started to serve the request.

**max_replicas[default_value=1]**: The maximum number of replicas for the deployment. Ray Serve Autoscaling will rely on the Ray Autoscaler to scale up more nodes when the currently available cluster resources (CPUs, GPUs, etc.) are not enough to support more replicas.

**target_num_ongoing_requests_per_replica[default_value=1]**: How many ongoing requests are expected to run concurrently per replica. The autoscaler scales up if the value is lower than the current number of ongoing requests per replica. Similarly, the autoscaler scales down if it's higher than the current number of ongoing requests. Scaling happens quicker if there's a high disparity between this value and the current number of ongoing requests.

> **downscale_delay_s[default_value=600.0]**: How long the cluster needs to wait before scaling down replicas. This controls the frequency of downscale decisions. For example, if your application takes a long time to do initialization work, you can increase `downscale_delay_s` to make the downscaling happen slowly.

Reviewer 3: ```suggestion
**downscale_delay_s[default_value=600.0]**: How long the cluster needs to wait before scaling down replicas. This controls the frequency of downscale decisions. For example, if your application initializes slowly, you can increase `downscale_delay_s` to make downscaling happen more infrequently and avoid reinitialization when the application needs to upscale again in the future.
```

* **`target_num_ongoing_requests_per_replica`** is the average number of ongoing requests per replica that the Serve autoscaler will try to ensure. Set this to a reasonable number (for example, 5) and adjust it based on your request processing length (the longer the requests, the smaller this number should be) as well as your latency objective (the shorter you want your latency to be, the smaller this number should be).
* **`min_replicas`** is the minimum number of replicas for the deployment. Set this to 0 if there are long periods of no traffic and some extra tail latency during upscale is acceptable. Otherwise, set this to what you think you need for low-traffic.
* **`max_replicas`** is the minimum number of replicas for the deployment. Set this to ~20% higher than what you think you need for peak traffic.

> How to set other relevant values in deployment config:

Reviewer 5: The repetition seems weird here. Can we just give a code snippet example to make it clear and merge the text?

Author: Done!
